src/maps/cave_map.rs

```diff
@@ -0,0 +1,276 @@
+use super::map_base::MapBase;
+use rand::{distributions::WeightedIndex, prelude::*, Rng};
+use std::cmp;
+
+pub struct CaveMap {
+    pub map: MapBase,
+    edge_repulsion_dist: u32,
+    //min_treasures: u32,
+    //max_treasures: u32,
+    coverage: f32,
+}
+impl CaveMap {
+    pub fn new(width: u32, height: u32) -> CaveMap {
+        let mut ret = CaveMap {
+            map: MapBase::new(width, height),
+            edge_repulsion_dist: 10,
+            coverage: 0.3,
+        };
+        ret.gen();
+        ret
+    }

```

Comment:
ok trying to remember all of my points from that comment that was deleted... and "adding" this immediately

Reply:
How do you make associated constants? That seems like the better option to me.
What's a free function?
I don't understand why calling gen inside new is an issue. If you always call gen after creating the CaveMap, as it is intended to be used, then why not put it inside new itself?

Reply:
Actually, I'll just look up free functions and associated constants on my own.

Reply:
The problem with .gen being in new is that is requires the user of the API to run gen every time they create a CaveMap. But the data that gen is writing to is public. Which means the user is allowed to have any valid MapBase in that memory (and may want to take advantage of that). So say they want to create a CaveMap that uses custom generation logic. They need run all of .gen... then immediately throw away that work and run the generation logic they actually wanted. Or say they want to load a CaveMap from disk (as part of a savegame or w/e). They instantiate a Cavemap, which runs its entire generation algo, then immediately discard that work and replace .map with the data from disk.

A free function is a non-method, non-associated function. ie, not part of an impl.

An associated constant is defined like so:

```
struct Structure {
        const CONST_NAME: i32 = 5;
}

Reply:
> is requires the user of the API to run gen every time they create a CaveMap
That is the intention yes. Yes, the user could then turn around and rewrite all the data, but that would be pointless. A CaveMap is basically *defined* by having a gen function different from other maps. If you can create a CaveMap that doesn't run it's gen function, then it's no different from a generic Map. So then maybe gen() should just be caveGen() and moved into Map, getting rid of CaveMap as a struct altogether. I'd be fine with this if there was someway to still break up the code across files, but I'm not sure if that's possible, and I don't want 30 different 200 line gen functions in the same file. As for loading a CaveMap, I'd say loading a map should just be a separate function.

Reply:
Ultimately, there is not much reason to tracking the initialization method (eg, .gen()) of a map via the type system (the CaveMap wrapper type around BaseMap) if all of the results of that method can be overwritten by API consumers (because map is pub).

If the intent is for CaveMap to *always* have a map state that was created by .gen, then I think the best solution is to:

* Make map private, and `impl AsRef<BaseMap>` for CaveMap, so users can extract the underlying map data and use it as they please, while still guaranteeing that a CaveMap is a CaveMap. This also lets them (almost) painless use CaveMaps where Basemaps are expected as inputs. And also do the associated constants thing.

Reply:
Actually, implementing AsRef is probably unnecessary.